File: algorithms/Longest ZigZag Path in a Binary Tree/Longest ZigZag Path in a Binary Tree.py
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        
        def helper(root):
            if root is None:
                return -1, -1
            
            leftRight = helper(root.left)[1] + 1
            rightLeft = helper(root.right)[0] + 1
            self.res = max(self.res, leftRight, rightLeft)
            return leftRight, rightLeft
        
        helper(root)
        return self.res

File: algorithms/L/Longest ZigZag Path in a Binary Tree/Longest ZigZag Path in a Binary Tree.cpp
class Solution {
    typedef long long ll;
    typedef pair<ll, ll> pi;
public:
    ll ans = 0;
    pi func(TreeNode* nd) {
        if(!nd){
            return {-1,-1};
        }
        pi p = { func(nd->left).second + 1, func(nd->right).first + 1 };
        ans = max({ans, p.first, p.second});
        return p;
    }
    int longestZigZag(TreeNode* root) {
        func(root);
        return ans;
    }
};
